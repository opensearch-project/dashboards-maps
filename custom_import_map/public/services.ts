/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CoreStart } from '../../../src/core/public';

export const postGeojson = async (requestBody: any, http: CoreStart['http']) => {
  try {
    const response = await http.post('../api/custom_import_map/_upload', {
      body: requestBody,
    });
    return response;
  } catch (e) {
    return e;
  }
};

export const getIndex = async (indexName: string, http: CoreStart['http']) => {
  try {
    const response = await http.post('../api/custom_import_map/_indices', {
      body: JSON.stringify({
        index: indexName,
      }),
    });
    return response;
  } catch (e) {
    return e;
  }
};
