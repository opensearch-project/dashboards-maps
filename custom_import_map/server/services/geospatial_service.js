/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export default class GeospatialService {
  constructor(driver) {
    this.driver = driver;
  }

  uploadGeojson = async (context, req, res) => {
    try {
      const { callAsCurrentUser } = await this.driver.asScoped(req);
      const uploadResponse = await callAsCurrentUser('geospatial.geospatialQuery', {
        body: req.body,
      });
      return res.ok({
        body: {
          ok: true,
          resp: uploadResponse,
        },
      });
    } catch (err) {
      return res.ok({
        body: {
          ok: false,
          resp: err.message,
        },
      });
    }
  };
}
