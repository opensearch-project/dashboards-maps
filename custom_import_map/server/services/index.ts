/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import GeospatialService from './geospatial_service';
import OpensearchService from './opensearch_service';

export { GeospatialService, OpensearchService };
