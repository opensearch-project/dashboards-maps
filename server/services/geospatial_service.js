/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export default class GeospatialService {
  constructor(driver) {
    this.driver = driver;
  }

  uploadGeojson = async (context, req, res) => {
    const dataSourceRefId = req.query.dataSourceId;
    let uploadResponse;
    try {
      if (dataSourceRefId) {
        const remoteDataSourceClient = context.dataSource.opensearch.legacy.getClient(
          dataSourceRefId
        ).callAPI;
        uploadResponse = await remoteDataSourceClient('geospatial.geospatialQuery', {
          body: req.body,
        });
      } else {
        const { callAsCurrentUser } = await this.driver.asScoped(req);
        uploadResponse = await callAsCurrentUser('geospatial.geospatialQuery', {
          body: req.body,
        });
      }

      return res.ok({
        body: {
          ok: true,
          resp: uploadResponse,
        },
      });
    } catch (err) {
      return res.ok({
        body: {
          ok: false,
          resp: err.message,
        },
      });
    }
  };
}
