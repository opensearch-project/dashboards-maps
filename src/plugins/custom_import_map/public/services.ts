/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const postGeojson = async (requestBody: any, http: any) => {
  try {
    const response = await http.post('../api/custom_import_map/_upload', {
      body: requestBody,
    });
    return response;
  } catch (e) {
    return e;
  }
};

export const getIndex = async (indexName: string, http: any) => {
  try {
    const response = await http.post('../api/custom_import_map/_indices', {
      body: JSON.stringify({
        index: indexName,
      }),
    });
    return response;
  } catch (e) {
    return e;
  }
};

export const getPlugins = async (http: any) => {
  try {
    const response = await http.post('../api/custom_import_map/_plugins', {});
    return response;
  } catch (e) {
    return e;
  }
};
