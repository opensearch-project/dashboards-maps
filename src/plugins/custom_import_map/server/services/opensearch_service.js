/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export default class OpensearchService {
  constructor(driver) {
    this.driver = driver;
  }

  getIndex = async (context, req, res) => {
    try {
      const { index } = req.body;
      const { callAsCurrentUser } = this.driver.asScoped(req);
      const indices = await callAsCurrentUser('cat.indices', {
        index,
        format: 'json',
        h: 'health,index,status',
      });
      return res.ok({
        body: {
          ok: true,
          resp: indices,
        },
      });
    } catch (err) {
      // Opensearch throws an index_not_found_exception which we'll treat as a success
      if (err.statusCode === 404) {
        return res.ok({
          body: {
            ok: false,
            resp: [],
          },
        });
      } else {
        return res.ok({
          body: {
            ok: false,
            resp: err.message,
          },
        });
      }
    }
  };

  getPlugins = async (context, req, res) => {
    try {
      const { callAsCurrentUser } = this.driver.asScoped(req);
      const plugins = await callAsCurrentUser('cat.plugins', {
        format: 'json',
      });
      return res.ok({
        body: {
          ok: true,
          resp: plugins,
        },
      });
    } catch (err) {
      return res.ok({
        body: {
          ok: false,
          resp: err.message,
        },
      });
    }
  };
}
