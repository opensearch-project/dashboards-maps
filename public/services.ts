/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CoreStart } from '../../../src/core/public';
import { createGetterSetter } from '../../../src/plugins/opensearch_dashboards_utils/common';
import { TimefilterContract } from '../../../src/plugins/data/public';

export const postGeojson = async (
  requestBody: any,
  http: CoreStart['http'],
  dataSourceRefId: string
) => {
  try {
    const query = dataSourceRefId ? { dataSourceId: dataSourceRefId } : undefined;

    return await http.post('../api/custom_import_map/_upload', {
      body: requestBody,
      ...(query && { query }),
    });
  } catch (e) {
    return e;
  }
};

export const getIndex = async (
  indexName: string,
  http: CoreStart['http'],
  dataSourceRefId: string
) => {
  try {
    const query = dataSourceRefId ? { dataSourceId: dataSourceRefId } : undefined;
    return await http.post('../api/custom_import_map/_indices', {
      body: JSON.stringify({ index: indexName }),
      ...(query && { query }),
    });
  } catch (e) {
    return e;
  }
};

export const [getTimeFilter, setTimeFilter] = createGetterSetter<TimefilterContract>('TimeFilter');
