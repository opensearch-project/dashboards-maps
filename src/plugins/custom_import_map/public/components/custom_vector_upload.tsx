/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect } from 'react';
import { EuiButton } from '@elastic/eui';
import { getPlugins } from '../services';
import { EmptyPrompt } from './empty_prompt';
import { VectorUploadOptions } from './vector_upload_options';
import { RegionMapOptionsProps } from '../../../../src/plugins/region_map/public';

function CustomVectorUpload(props: RegionMapOptionsProps) {
  const GEOSPATIAL_PLUGIN = 'opensearch-geospatial';
  const [isGeospatialPluginInstalled, setIsGespatialPluginInstalled] = React.useState(false);

  // Logic for implementing empty state pattern
  // In case geospatial plugin is not installed, an empty prompt with details is shown.
  // TODO: Remove the plugin check logic once we merge goal 2 (visualizing custom map) 
  // related changes in OpenSeatch Dashboards
  useEffect(() => {
    const checkIfGeospatialPluginIsInstalled = async () => {
      const result = await getPlugins(props.vis.http);
      const installedPlugins = result.resp.map((plugin) => {
        return plugin.component;
      });
      if (installedPlugins.includes(GEOSPATIAL_PLUGIN)) {
        return true;
      }

      return false;
    };

    async function getPluginInfo() {
      const installedFlag = await Promise.resolve(checkIfGeospatialPluginIsInstalled());
      if (installedFlag) {
        setIsGespatialPluginInstalled(true);
      } else {
        setIsGespatialPluginInstalled(false);
      }
    }
    getPluginInfo();
  }, [props]);

  const emptyPromptProps = {
    iconType: 'popout',
    title: 'Missing geospatial plugin',
    bodyFragment: 'Install the geospatial plugin in order to upload custom vector maps.',
    actions: (
      <EuiButton href="https://opensearch.org/docs/latest/" color="primary" fill>
        Learn more
      </EuiButton>
    ),
  };

  return (
    <div>
      {isGeospatialPluginInstalled ? (
        <VectorUploadOptions {...props} />
      ) : (
        <EmptyPrompt {...emptyPromptProps} />
      )}
    </div>
  );
}

export { CustomVectorUpload };
