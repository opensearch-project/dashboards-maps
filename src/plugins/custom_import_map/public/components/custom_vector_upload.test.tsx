/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import renderer, { act } from 'react-test-renderer';
import * as React from 'react';
import { CustomVectorUpload } from './custom_vector_upload';
import * as serviceApiCalls from '../services';

describe('custom vector upload component', () => {
  const mockDataWhenPluginIsUninstalled = {
    ok: true,
    resp: [
      {
        name: 'integTest-0',
        component: 'opensearch',
        version: '2.1.0.0-SNAPSHOT',
      },
    ],
  };
  const mockDataWhenPluginIsInstalled = {
    ok: true,
    resp: [
      {
        name: 'integTest-0',
        component: 'opensearch-geospatial',
        version: '2.1.0.0-SNAPSHOT',
      },
    ],
  };
  const props = {
    vis: {
      http: '',
      notifications: '',
    },
  };

  it('renders the empty prompt if geospatial plugin is not installed', async () => {
    jest.spyOn(serviceApiCalls, 'getPlugins').mockImplementation((http) => {
      return mockDataWhenPluginIsUninstalled;
    });

    let tree;
    await act(async () => {
      tree = renderer.create(<CustomVectorUpload {...props} />);
    });
    expect(tree.toJSON().children[0].props.testId).toBe('empty-prompt-id');
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders the upload customer div if geospatial plugin is installed', async () => {
    jest.spyOn(serviceApiCalls, 'getPlugins').mockImplementation((http) => {
      return mockDataWhenPluginIsInstalled;
    });

    let tree;
    await act(async () => {
      tree = renderer.create(<CustomVectorUpload {...props} />);
    });
    expect(tree.toJSON().children[0].props.id).toBe('uploadCustomVectorMap');
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
