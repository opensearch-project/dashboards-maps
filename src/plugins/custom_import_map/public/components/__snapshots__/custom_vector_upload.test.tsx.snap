// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom vector upload component renders the empty prompt if geospatial plugin is not installed 1`] = `
<div>
  <div
    className="euiEmptyPrompt"
    testId="empty-prompt-id"
  >
    <div>
      EuiIconMock
    </div>
    <div
      className="euiSpacer euiSpacer--m"
    />
    <h2
      className="euiTitle euiTitle--medium"
    >
      Missing geospatial plugin
    </h2>
    <span
      className="euiTextColor euiTextColor--subdued"
    >
      <div
        className="euiSpacer euiSpacer--m"
      />
      <div
        className="euiText euiText--medium"
      >
        <p>
          Install the geospatial plugin in order to upload custom vector maps.
        </p>
      </div>
    </span>
    <div
      className="euiSpacer euiSpacer--l"
    />
    <a
      className="euiButton euiButton--primary euiButton--fill"
      disabled={false}
      href="https://opensearch.org/docs/latest/"
      rel="noreferrer"
      style={
        Object {
          "minWidth": undefined,
        }
      }
    >
      <span
        className="euiButtonContent euiButton__content"
      >
        <span
          className="euiButton__text"
        >
          Learn more
        </span>
      </span>
    </a>
  </div>
</div>
`;

exports[`custom vector upload component renders the upload customer div if geospatial plugin is installed 1`] = `
<div>
  <div
    id="uploadCustomVectorMap"
  >
    <div
      aria-label="import-vector-map-card"
      className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiCard euiCard--leftAligned"
    >
      <div
        className="euiCard__content"
      >
        <span
          className="euiTitle euiTitle--small euiCard__title"
          id="generated-idTitle"
        >
          
        </span>
        
        <div
          className="euiCard__children"
        >
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--s"
          />
          <div
            aria-label="upload-map-text"
            className="euiText euiText--small"
          >
            Upload map
          </div>
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--m"
          />
          <div
            aria-label="form-row-for-file-picker"
            className="euiFormRow"
            id="generated-id-row"
          >
            <div
              className="euiFormRow__fieldWrapper"
            >
              <div
                className="euiFilePicker euiFilePicker--large"
              >
                <div
                  className="euiFilePicker__wrap"
                >
                  <input
                    accept=".json,.geojson"
                    aria-describedby="generated-id-filePicker__prompt"
                    aria-label="geojson-file-picker"
                    className="euiFilePicker__input"
                    data-test-subj="filePicker"
                    data-testid="filePicker"
                    id="generated-id"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDrop={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="file"
                  />
                  <div
                    className="euiFilePicker__prompt"
                    id="generated-id-filePicker__prompt"
                  >
                    <div>
                      EuiIconMock
                    </div>
                    <div
                      className="euiFilePicker__promptText"
                    >
                      Select or drag and drop a json file
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--m"
          />
          <div
            aria-label="geojson-file-format-text"
            className="euiText euiText--extraSmall"
          >
            <div
              className="euiTextColor euiTextColor--subdued"
            >
              <p>
                <span
                  className="formattedSpan"
                >
                  Formats accepted: .json, .geojson
                </span>
                <span
                  className="formattedSpan"
                >
                  Max size: 25 MB
                </span>
                <span
                  className="formattedSpan"
                >
                  Coordinates must be in EPSG:4326 coordinate reference system.
                </span>
              </p>
            </div>
          </div>
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--m"
          />
          <div
            aria-label="map-name-prefix-text"
            className="euiText euiText--small"
          >
            Map name prefix
          </div>
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--m"
          />
          <div
            className="euiFormControlLayout"
          >
            <div
              className="euiFormControlLayout__childrenWrapper"
            >
              <input
                aria-label="map-name-text-field"
                className="euiFieldText"
                data-test-subj="customIndex"
                data-testid="customIndex"
                id="customIndex"
                label="Map name"
                name="customIndex"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Enter a valid map name prefix"
                required={true}
                tabIndex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--m"
          />
          <div
            aria-label="map-name-guidelines-text"
            className="euiText euiText--extraSmall"
          >
            <div
              className="euiTextColor euiTextColor--subdued"
            >
              Map name guidleines:
              <ul>
                <li>
                   Map name prefix must contain 1-250 characters. 
                </li>
                <li>
                   Map name prefix must start with a-z.
                </li>
                <li>
                   Valid characters are a-z, 0-9, - and _ .
                </li>
                <li>
                  The final map name will be the entered prefix here followed by -map as the suffix.
                </li>
              </ul>
            </div>
          </div>
          <div
            aria-label="map-name-error-text"
            className="euiText euiText--extraSmall"
          >
            <span
              aria-label="map-name-error-text-color"
              className="euiTextColor euiTextColor--danger"
            >
              <p
                name="errorIndexName"
              />
            </span>
          </div>
          <div
            aria-label="medium-spacer"
            className="euiSpacer euiSpacer--m"
          />
          <div
            className="importFileButton"
          >
            <button
              aria-label="import-file-button"
              className="euiButton euiButton--primary euiButton--fill"
              disabled={false}
              id="submitButton"
              onClick={[Function]}
              style={
                Object {
                  "minWidth": undefined,
                }
              }
              type="submit"
            >
              <span
                className="euiButtonContent euiButton__content"
              >
                <span
                  className="euiButton__text"
                >
                  Import file
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
